bin/varnishd/cache/cache_ban_lurker.c

```diff
@@ -169,6 +169,21 @@ ban_lurker_getfirst(struct vsl_log *vsl, struct ban *bt)
 		if (!Lck_Trylock(&oh->mtx)) {
 			if (oc->refcnt == 0 || oc->flags & OC_F_BUSY) {
 				Lck_Unlock(&oh->mtx);
+			} else if (oc->flags & (OC_F_DYING | OC_F_FAILED)) {
+				/*
+				 * For a dying object we seize the opportunity
+				 * to remove it completely off the ban list,
+				 * now that we have both the oh and ban
+				 * mutexes.
+				 */
+				AZ(oc->flags & OC_F_BUSY);
+				noc = VTAILQ_NEXT(oc, ban_list);
+				VTAILQ_REMOVE(&bt->objcore, oc, ban_list);
+				oc->ban = NULL;
+				bt->refcount--;
+				Lck_Unlock(&oh->mtx);
+				oc = noc;
+				continue;
```

Comment:
This does look like a good idea, but if we add this, we should also optimize https://github.com/varnishcache/varnish-cache/blob/master/bin/varnishd/cache/cache_ban.c#L265 to not take the `ban_mtx` for `oc->ban == NULL` (and re-check under the lock)

Reply:
I think you are right, but we should really look for potential races if we do this. Martin had the same idea, but was not completely sure it was safe.

Reply:
I did have my doubts as well and did check, so far I think it is safe because once we have a ban, we only ever modify it and never transition from `NULL` to `!NULL` except in `BAN_RefBan()` and `BAN_NewObjCore()`. So I think that if we re-check under the `ban_mtx`, it's safe.